app/src/org/commcare/tts/TextToSpeechConverter.kt

```diff
@@ -0,0 +1,208 @@
+        // The callbacks specified here can be called from multiple threads.
+        override fun onDone(utteranceId: String?) { }
+
+        override fun onError(utteranceId: String?) {
+            Handler(Looper.getMainLooper()).post {
+                mTTSCallback?.speakFailed()
+            }
+        }
+
+        override fun onStart(utteranceId: String?) { }
+    }
+
+    /**
+     * Adds a listener to register callbacks for different states of Text-To-Speech engine.
+     */
+    fun setListener(listener: TextToSpeechCallback) {
+        mTTSCallback = listener
+    }
+
+    /**
+     * Attempts to speak the specified text.
+     */
+    fun speak(context: Context, text: String) {
+        if (!mInitialized) {
+            initialize(context, text)
+            return
+        }
+        // Handle empty text
+        if (TextUtils.isEmpty(text)) {
+            return
+        }
+        mTextToSpeech?.let { tts ->
+            if (text.length > MAX_TEXT_LENGTH) {
+                text.chunked(MAX_TEXT_LENGTH).forEach {
+                    speakInternal(tts, it, TextToSpeech.QUEUE_ADD)
+                }
+            } else {
+                speakInternal(tts, text)
+            }
+        } ?: run {
+            mTTSCallback?.initFailed()
+        }
+    }
+
+    /**
+     * Attempts to stop the TTS.
+     */
+    fun stop() {
+        mTextToSpeech?.let {
+            it.stop()
+        }
+    }
+
+    /**
+     * Attempts to shutdown the TTS engine. No calls should be made to this object after calling this method.
+     * Good to call this from onDestroy().
+     */
+    fun shutDown() {
+        mTextToSpeech?.let {
+            it.shutdown()
+        }
+        mInitialized = false
+    }
+
+    /**
+     * Changes the locale of text to speech engine.
+     */
+    fun changeLocale(language: String) {
+        mTextToSpeech?.let { tts ->
+            setLocale(tts, LinkedList(listOf(Locale(language, Locale.getDefault().country))))
+        }
+    }
+
+    private fun speakInternal(tts: TextToSpeech, text: String, queueMode: Int = TextToSpeech.QUEUE_FLUSH) {
+        val utteranceId = System.currentTimeMillis().toString()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            tts.speak(text, queueMode, null, utteranceId)
+        } else {
+            val params = HashMap<String, String>()
+            params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = utteranceId
+            tts.speak(text, queueMode, params)
+        }
+    }
+
+    private fun isTextLong(text: String): Boolean {
+        // TTS can only speak 4000 characters at max at a time.
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            text.length > TextToSpeech.getMaxSpeechInputLength()
+        } else {
+            text.length > MAX_TEXT_LENGTH
+        }
+    }
+
+    /**
+     * Initializes the Text-To-Speech engine
+     */
+    private fun initialize(context: Context, text: String) {
```

Comment:
Should not this called again after user change a language in CommCare ?

Reply:
We only need to change the language which is done here https://github.com/dimagi/commcare-android/pull/2411/files#diff-c7cb0f5ab8dcc60f97dfaf06c122ffa5e54311f12db10e5bd388d3d4035feffbR825

Reply:
Ahh I missed that. thanks